ext/nnstreamer/tensor_filter/tensor_filter_llamacpp.cc

```diff
@@ -286,26 +327,40 @@ TensorFilterLlamaCpp::invoke_dynamic (GstTensorFilterProperties *prop,
       if (n < 0) {
         throw std::invalid_argument ("Failed to convert token to piece");
       }
-      std::string s (buf, n);
-      result += s;
 
-      /* prepare the next batch with the sampled token */
+      output = g_new0 (GstTensorMemory, 1);
+      createOutputTensor (buf, n, output, prop);
+      /* nnstreamer frees GstTensorMemory output after use. Do not free it. */
+      nnstreamer_filter_dispatch_output_async (prop, output);
+
       batch = llama_batch_get_one (&new_token_id, 1);
     }
   }
+}
 
-  /* set output info */
-  output[0].size = result.length ();
-  output[0].data = strndup (result.c_str (), result.length ());
+/**
+ * @brief Invoke llamacpp using input tensors
+ */
+void
+TensorFilterLlamaCpp::invoke_dynamic (GstTensorFilterProperties *prop,
+    const GstTensorMemory *input, GstTensorMemory *output)
+{
+  UNUSED (output);
+  std::string prompt;
 
-  gst_tensors_info_free (&prop->output_meta);
-  prop->output_meta.num_tensors = 1;
-  prop->output_meta.format = _NNS_TENSOR_FORMAT_FLEXIBLE;
-  prop->output_meta.info[0].type = _NNS_UINT8;
-  prop->output_meta.info[0].dimension[0] = output[0].size;
+  if (!input || !input[0].data || input[0].size == 0) {
+    throw std::invalid_argument ("Invalid input tensor data");
+  }
+  if (output_thread.joinable ()) {
```

Comment:
Should we create new thread for each input?
I suggest adding message queue with lock, pushing message (includes input data),  and fetching a message in the thread.

Reply:
I've updated it, thank you.